// moq.ts/esm2015/lib/expression.equality-comparers/expression.equality-comparer.js
import { It } from "../reflector/expression-predicates";
import { GetPropertyExpression, InOperatorExpression, FunctionExpression, MethodExpression, NewOperatorExpression, SetPropertyExpression } from "../reflector/expressions";
/**
 * @hidden
 */
export class ExpressionEqualityComparer {
    constructor(getPropertyExpressionMatcher, setPropertyExpressionMatcher, methodExpressionMatcher, namedMethodExpressionMatcher, inOperatorExpressionMatcher, newOperatorExpressionMatcher, itEqualityComparer) {
        this.getPropertyExpressionMatcher = getPropertyExpressionMatcher;
        this.setPropertyExpressionMatcher = setPropertyExpressionMatcher;
        this.methodExpressionMatcher = methodExpressionMatcher;
        this.namedMethodExpressionMatcher = namedMethodExpressionMatcher;
        this.inOperatorExpressionMatcher = inOperatorExpressionMatcher;
        this.newOperatorExpressionMatcher = newOperatorExpressionMatcher;
        this.itEqualityComparer = itEqualityComparer;
    }
    equals(left, right) {
        if (left === right)
            return true;
        if (left instanceof It || right instanceof It) {
            return this.itEqualityComparer.equals(left, right);
        }
        if (left instanceof GetPropertyExpression && right instanceof GetPropertyExpression) {
            return this.getPropertyExpressionMatcher.equals(left, right);
        }
        if (left instanceof SetPropertyExpression && right instanceof SetPropertyExpression) {
            return this.setPropertyExpressionMatcher.equals(left, right);
        }
        if (left instanceof InOperatorExpression && right instanceof InOperatorExpression) {
            return this.inOperatorExpressionMatcher.equals(left, right);
        }
        if (left instanceof FunctionExpression && right instanceof FunctionExpression) {
            return this.methodExpressionMatcher.equals(left, right);
        }
        if (left instanceof MethodExpression && right instanceof MethodExpression) {
            return this.namedMethodExpressionMatcher.equals(left, right);
        }
        if (left instanceof NewOperatorExpression && right instanceof NewOperatorExpression) {
            return this.newOperatorExpressionMatcher.equals(left, right);
        }
        return false;
    }
}
