// moq.ts/esm2015/lib/auto-mocking/name-formatters/auto-mock-name.formatter.js
import { GetPropertyExpression, FunctionExpression, MethodExpression, NewOperatorExpression } from "../../reflector/expressions";
/**
 * @hidden
 */
export class AutoMockNameFormatter {
    constructor(namePrefixProvider, methodFormatter, propertyKeyFormatter, namedMethodFormatter, constantFormatter) {
        this.namePrefixProvider = namePrefixProvider;
        this.methodFormatter = methodFormatter;
        this.propertyKeyFormatter = propertyKeyFormatter;
        this.namedMethodFormatter = namedMethodFormatter;
        this.constantFormatter = constantFormatter;
    }
    format(name, expression) {
        const prefix = this.namePrefixProvider.get(name);
        if (expression instanceof FunctionExpression) {
            return `${prefix}${this.methodFormatter.format(expression)}`;
        }
        if (expression instanceof GetPropertyExpression) {
            return `${prefix}.${this.propertyKeyFormatter.format(expression.name)}`;
        }
        if (expression instanceof MethodExpression) {
            return `${prefix}.${this.namedMethodFormatter.format(expression)}`;
        }
        if (expression instanceof NewOperatorExpression) {
            return `new ${name}(${this.constantFormatter.format(expression.args)})`;
        }
        return `${name}[${expression}]`;
    }
}
