// moq.ts/esm2015/lib/presets/preset-builder.js
import { PlayTimes } from "../playables/play-times";
/**
 * The default implementation of {@link IPresetBuilder} interface.
 * Is it not intended to be used outside of the moq library.
 *
 * @hidden
 */
export class PresetBuilder {
    constructor(returnsPresetFactory, throwsPresetFactory, mimicsPresetFactory, callbackPresetFactory, returnsAsyncPresetFactory, throwsAsyncPresetFactory, target, playable = PlayTimes.Always()) {
        this.returnsPresetFactory = returnsPresetFactory;
        this.throwsPresetFactory = throwsPresetFactory;
        this.mimicsPresetFactory = mimicsPresetFactory;
        this.callbackPresetFactory = callbackPresetFactory;
        this.returnsAsyncPresetFactory = returnsAsyncPresetFactory;
        this.throwsAsyncPresetFactory = throwsAsyncPresetFactory;
        this.target = target;
        this.playable = playable;
    }
    returnsAsync(value) {
        return this.returnsAsyncPresetFactory(this.target, this.playable, value);
    }
    throwsAsync(exception) {
        return this.throwsAsyncPresetFactory(this.target, this.playable, exception);
    }
    mimics(origin) {
        return this.mimicsPresetFactory(this.target, this.playable, origin);
    }
    returns(value) {
        return this.returnsPresetFactory(this.target, this.playable, value);
    }
    throws(exception) {
        return this.throwsPresetFactory(this.target, this.playable, exception);
    }
    callback(callback) {
        return this.callbackPresetFactory(this.target, this.playable, callback);
    }
    play(playable) {
        this.playable = playable;
        return this;
    }
}
