// moq.ts/esm2015/lib/core/mock-core.js
/**
 * The internal core of {@link Mock} class.
 */
export class MockCore {
    constructor(options, tracker, injector, reflector, interceptor, verifier, prototypeStorage, mock, setupFactory) {
        this.options = options;
        this.tracker = tracker;
        this.injector = injector;
        this.reflector = reflector;
        this.interceptor = interceptor;
        this.verifier = verifier;
        this.prototypeStorage = prototypeStorage;
        this.mock = mock;
        this.setupFactory = setupFactory;
    }
    get name() {
        return this.options.name;
    }
    resolve(token) {
        return this.injector.get(token, null);
    }
    setup(expression) {
        const expressions = this.reflector.reflect(expression);
        return this.setupFactory.create(expressions);
    }
    verify(expression, times) {
        const expressions = this.reflector.reflect(expression);
        this.verifier.test(expressions, times);
        return this.mock;
    }
    object() {
        return this.interceptor.object();
    }
    prototypeof(prototype) {
        this.prototypeStorage.set(prototype);
        return this.mock;
    }
    /**
     * @experimental
     */
    insequence(sequence, expression) {
        sequence.add(this, expression);
        return this.mock;
    }
}
