// moq.ts/esm2015/lib/interceptors/has.trap.js
import { InOperatorExpression } from "../reflector/expressions";
/**
 * @hidden
 */
export class HasTrap {
    constructor(tracker, propertiesValueStorage, interactionPlayer, inOperatorInteractionExplorer, hasPropertyExplorer, hasMethodExplorer, presetPlayablesUpdater) {
        this.tracker = tracker;
        this.propertiesValueStorage = propertiesValueStorage;
        this.interactionPlayer = interactionPlayer;
        this.inOperatorInteractionExplorer = inOperatorInteractionExplorer;
        this.hasPropertyExplorer = hasPropertyExplorer;
        this.hasMethodExplorer = hasMethodExplorer;
        this.presetPlayablesUpdater = presetPlayablesUpdater;
    }
    intercept(property) {
        const interaction = new InOperatorExpression(property);
        this.tracker.add(interaction);
        if (this.propertiesValueStorage.has(property)) {
            return true;
        }
        if (this.inOperatorInteractionExplorer.has(property)) {
            return this.interactionPlayer.play(interaction);
        }
        this.presetPlayablesUpdater.update(interaction, undefined);
        if (this.hasPropertyExplorer.has(property)) {
            return true;
        }
        if (this.hasMethodExplorer.has(property)) {
            return true;
        }
        return false;
    }
}
