// moq.ts/esm2015/lib/interceptors/proxy.factory.js
/**
 * @hidden
 */
export class ProxyFactory {
    constructor(options, getTrap, setTrap, hasTrap, applyTrap, getPrototypeOfTrap, setPrototypeOfTrap, constructTrap) {
        this.options = options;
        this.getTrap = getTrap;
        this.setTrap = setTrap;
        this.hasTrap = hasTrap;
        this.applyTrap = applyTrap;
        this.getPrototypeOfTrap = getPrototypeOfTrap;
        this.setPrototypeOfTrap = setPrototypeOfTrap;
        this.constructTrap = constructTrap;
    }
    object() {
        if (this._proxy === undefined) {
            this._proxy = this.createObject();
        }
        return this._proxy;
    }
    createObject() {
        const options = {
            get: (target, name) => this.getTrap.intercept(name),
            set: (target, name, value) => this.setTrap.intercept(target, name, value),
            has: (target, name) => this.hasTrap.intercept(name),
            apply: (target, thisArg, args) => this.applyTrap.intercept(target, thisArg, args),
            getPrototypeOf: () => this.getPrototypeOfTrap.intercept(),
            setPrototypeOf: (target, prototype) => this.setPrototypeOfTrap.intercept(prototype),
            construct: (target, args) => this.constructTrap.intercept(args)
        };
        if (this.options.name) {
            options["mockName"] = this.options.name;
        }
        return new Proxy(this.options.target, options);
    }
}
