// moq.ts/esm2015/lib/interceptors/set.trap.js
import { SetPropertyExpression } from "../reflector/expressions";
import { MoqAPI } from "../moq";
/**
 * @hidden
 */
export class SetTrap {
    constructor(tracker, propertiesValueStorage, interactionPlayer, propertyIsReadOnlyTester) {
        this.tracker = tracker;
        this.propertiesValueStorage = propertiesValueStorage;
        this.interactionPlayer = interactionPlayer;
        this.propertyIsReadOnlyTester = propertyIsReadOnlyTester;
    }
    intercept(target, property, value) {
        const expression = new SetPropertyExpression(property, value);
        this.tracker.add(expression);
        if (property === MoqAPI) {
            return false;
        }
        if (this.propertyIsReadOnlyTester.isReadOnly(property) === true) {
            return false;
        }
        const accepted = this.interactionPlayer.play(expression);
        if (accepted === true || accepted === undefined) {
            this.propertiesValueStorage.set(property, value);
        }
        return accepted === undefined ? true : accepted;
    }
}
