// moq.ts/esm2015/lib/reflector/expressions.js
export class Expression {
    constructor(name, args) {
        this.name = name;
        this.args = args;
    }
}
/**
 * This class represents an invocation of an instance method.
 * It provides access to the name of it and the parameters.
 */
export class MethodExpression extends Expression {
    constructor(name, args) {
        super(name, args);
    }
}
/**
 * This class represents an invocation of a function.
 * It provides access to the parameters.
 */
export class FunctionExpression extends Expression {
    constructor(args) {
        super(undefined, args);
    }
}
/**
 * This class represents a property accessing.
 * It provides access to the name of property.
 */
export class GetPropertyExpression extends Expression {
    constructor(name) {
        super(name, undefined);
    }
}
/**
 * This class represents a property write interaction.
 * It provides access to the name of property and the value.
 */
export class SetPropertyExpression extends Expression {
    constructor(name, value) {
        super(name, [value]);
        this.value = value;
    }
}
/**
 * This class represents applying of [in operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in).
 * It provides access to the name of property.
 */
export class InOperatorExpression extends Expression {
    constructor(name) {
        super(name, undefined);
    }
}
/**
 * This class represents applying of [new operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/new).
 * It provides access to the arguments of the constructor.
 */
export class NewOperatorExpression extends Expression {
    constructor(args) {
        super(undefined, args);
    }
}
