// moq.ts/esm2015/lib/auto-mocking/expression.guards/expression.formatters/expression.formatter.js
import { FunctionExpression, GetPropertyExpression, MethodExpression, NewOperatorExpression } from "../../../reflector/expressions";
export class ExpressionFormatter {
    constructor(functionFormatter, propertyKeyFormatter, instanceMethodFormatter, constantFormatter) {
        this.functionFormatter = functionFormatter;
        this.propertyKeyFormatter = propertyKeyFormatter;
        this.instanceMethodFormatter = instanceMethodFormatter;
        this.constantFormatter = constantFormatter;
    }
    format(expression, name) {
        if (expression instanceof FunctionExpression) {
            return this.functionFormatter.format(expression);
        }
        if (expression instanceof GetPropertyExpression) {
            return `.${this.propertyKeyFormatter.format(expression.name)}`;
        }
        if (expression instanceof MethodExpression) {
            return `.${this.instanceMethodFormatter.format(expression)}`;
        }
        if (expression instanceof NewOperatorExpression) {
            return `new ${name}(${this.constantFormatter.format(expression.args)})`;
        }
        return `[${expression}]`;
    }
}
