// moq.ts/esm2015/lib/auto-mocking/expression.guards/expression.validator.js
import { FunctionExpression, GetPropertyExpression, InOperatorExpression, MethodExpression, NewOperatorExpression, SetPropertyExpression } from "../../reflector/expressions";
import { It } from "../../reflector/expression-predicates";
export class ExpressionValidator {
    constructor(functionExpressionGuard, getPropertyExpressionGuard, setPropertyExpressionGuard, inOperatorExpressionGuard, instanceMethodExpressionGuard, newOperatorExpressionGuard) {
        this.functionExpressionGuard = functionExpressionGuard;
        this.getPropertyExpressionGuard = getPropertyExpressionGuard;
        this.setPropertyExpressionGuard = setPropertyExpressionGuard;
        this.inOperatorExpressionGuard = inOperatorExpressionGuard;
        this.instanceMethodExpressionGuard = instanceMethodExpressionGuard;
        this.newOperatorExpressionGuard = newOperatorExpressionGuard;
    }
    validate(expression) {
        switch (expression.constructor) {
            case FunctionExpression:
                return this.functionExpressionGuard.validate(expression);
            case GetPropertyExpression:
                return this.getPropertyExpressionGuard.validate(expression);
            case SetPropertyExpression:
                return this.setPropertyExpressionGuard.validate(expression);
            case InOperatorExpression:
                return this.inOperatorExpressionGuard.validate(expression);
            case MethodExpression:
                return this.instanceMethodExpressionGuard.validate(expression);
            case NewOperatorExpression:
                return this.newOperatorExpressionGuard.validate(expression);
            case It:
                return false;
            default:
                throw new Error(`Unknown expression type: ${expression.constructor.name}`);
        }
    }
}
