// moq.ts/esm2015/lib/formatters/interaction.formatter.js
import { GetPropertyExpression, InOperatorExpression, FunctionExpression, MethodExpression, NewOperatorExpression, SetPropertyExpression } from "../reflector/expressions";
import { It } from "../reflector/expression-predicates";
/**
 * @hidden
 */
export class InteractionFormatter {
    constructor(getPropertyFormatter, setPropertyFormatter, methodFormatter, namedMethodFormatter, constantFormatter, inOperatorFormatter, newOperatorFormatter) {
        this.getPropertyFormatter = getPropertyFormatter;
        this.setPropertyFormatter = setPropertyFormatter;
        this.methodFormatter = methodFormatter;
        this.namedMethodFormatter = namedMethodFormatter;
        this.constantFormatter = constantFormatter;
        this.inOperatorFormatter = inOperatorFormatter;
        this.newOperatorFormatter = newOperatorFormatter;
    }
    format(interaction) {
        if (interaction instanceof GetPropertyExpression)
            return this.getPropertyFormatter.format(interaction);
        if (interaction instanceof SetPropertyExpression)
            return this.setPropertyFormatter.format(interaction);
        if (interaction instanceof InOperatorExpression)
            return this.inOperatorFormatter.format(interaction);
        if (interaction instanceof FunctionExpression)
            return this.methodFormatter.format(interaction);
        if (interaction instanceof MethodExpression)
            return this.namedMethodFormatter.format(interaction);
        if (interaction instanceof NewOperatorExpression)
            return this.newOperatorFormatter.format(interaction);
        if (interaction instanceof It)
            return this.constantFormatter.format(interaction);
        return undefined;
    }
}
