// moq.ts/esm2015/lib/playables/play-times.js
import { PlayableExactly } from "./playable-exactly";
import { PlayableNever } from "./playable-never";
import { PlayableOnce } from "./playable-once";
import { PlayableAlways } from "./playable-always";
import { PlayableSequence } from "./playable-sequence";
/**
 * This class sets rules when a configured setup should be applied to the target invocation or not.
 *
 * @example
 * ```typescript
 *
 * const value = 'value';
 * const object = new Mock<Function>()
 * .setup(instance => instance(1))
 * .play(PlayTimes.Once())
 * .returns(value)
 * .object();
 *
 * expect(object(1).toBe(value);
 * expect(object(1).toBe(undefined);
 * ```
 */
export class PlayTimes {
    /**
     * The configured setup will be applied to invocations exactly n-times. After that it will be ignored.
     */
    static Exactly(count) {
        return new PlayableExactly(count);
    }
    /**
     * The configured setup will be always applied to invocations.
     */
    static Always() {
        return PlayTimes.always;
    }
    /**
     * The configured setup will be never applied to invocations.
     */
    static Never() {
        return PlayTimes.never;
    }
    /**
     * The configured setup will be applied only to the first invocation.
     */
    static Once() {
        return new PlayableOnce();
    }
    /**
     * The configured setup will be applied or not accordingly to the value in the sequence.
     *
     * @example
     * ```typescript
     *
     * const value = 'value';
     * const object = new Mock<Function>()
     * .setup(instance => instance(1))
     * .play(PlayTimes.Sequence([false, true]))
     * .returns(value)
     * .object();
     *
     * expect(object(1).toBe(undefined);
     * expect(object(1).toBe(value);
     * expect(object(1).toBe(undefined);
     * ```
     */
    static Sequence(sequence) {
        return new PlayableSequence(sequence);
    }
}
PlayTimes.always = new PlayableAlways();
PlayTimes.never = new PlayableNever();
